server/src/main/java/org/elasticsearch/ingest/IngestSourceAndMetadata.java

```diff
@@ -0,0 +1,613 @@
+        }
+        if (versionType != null) {
+            metadata.put(IngestDocument.Metadata.VERSION_TYPE.getFieldName(), VersionType.toString(versionType));
+        }
+        return metadata;
+    }
+
+    /**
+     * Returns a new metadata map and the existing source map with metadata removed.
+     */
+    public static Tuple<Map<String, Object>, Map<String, Object>> splitSourceAndMetadata(Map<String, Object> sourceAndMetadata) {
+        if (sourceAndMetadata instanceof IngestSourceAndMetadata ingestSourceAndMetadata) {
+            return new Tuple<>(new HashMap<>(ingestSourceAndMetadata.source), new HashMap<>(ingestSourceAndMetadata.metadata));
+        }
+        Map<String, Object> metadata = Maps.newHashMapWithExpectedSize(IngestDocument.Metadata.values().length);
+        Map<String, Object> source = new HashMap<>(sourceAndMetadata);
+        for (String metadataName : VALIDATORS.keySet()) {
+            if (sourceAndMetadata.containsKey(metadataName)) {
+                metadata.put(metadataName, source.remove(metadataName));
+            }
+        }
+        return new Tuple<>(source, metadata);
+    }
+
+    /**
+     * get the source map, if externally modified then the guarantees of this class are not enforced
+     */
+    public Map<String, Object> getSource() {
+        return source;
+    }
+
+    /**
+     * get the metadata map, if externally modified then the guarantees of this class are not enforced
+     */
+    public Map<String, Object> getMetadata() {
+        return metadata;
+    }
+
+    // These are available to scripts
+    public String getIndex() {
+        return getString(IngestDocument.Metadata.INDEX.getFieldName());
+    }
+
+    public void setIndex(String index) {
+        put(IngestDocument.Metadata.INDEX.getFieldName(), index);
+    }
+
+    public String getId() {
+        return getString(IngestDocument.Metadata.ID.getFieldName());
+    }
+
+    public void setId(String id) {
+        put(IngestDocument.Metadata.ID.getFieldName(), id);
+    }
+
+    public String getRouting() {
+        return getString(IngestDocument.Metadata.ROUTING.getFieldName());
+    }
+
+    public void setRouting(String routing) {
+        put(IngestDocument.Metadata.ROUTING.getFieldName(), routing);
+    }
+
+    public String getVersionType() {
+        return getString(IngestDocument.Metadata.VERSION_TYPE.getFieldName());
+    }
+
+    public void setVersionType(String versionType) {
+        put(IngestDocument.Metadata.VERSION_TYPE.getFieldName(), versionType);
+    }
+
+    public long getVersion() {
+        Number version = getNumber(IngestDocument.Metadata.VERSION.getFieldName());
+        assert version != null : IngestDocument.Metadata.VERSION.getFieldName() + " validation allowed null version";
+        return version.longValue();
+    }
+
+    public void setVersion(long version) {
+        put(IngestDocument.Metadata.VERSION.getFieldName(), version);
+    }
+
+    // timestamp isn't backed by the map
```

Comment:
why isn't timestamp part of the metadata map?

Reply:
It's a new field (https://github.com/elastic/elasticsearch/issues/80038) so I didn't think it was necessary to backport it into the ctx map and create a potential bwc issue.